src/StaticWebAssetsSdk/Tasks/DefineStaticWebAssets.Cache.cs

```diff
@@ -0,0 +1,217 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.AspNetCore.StaticWebAssets.Tasks.Utils;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using static Microsoft.AspNetCore.StaticWebAssets.Tasks.FingerprintPatternMatcher;
+
+namespace Microsoft.AspNetCore.StaticWebAssets.Tasks;
+
+public partial class DefineStaticWebAssets : Task
+{
+    private DefineStaticWebAssetsCache GetOrCreateAssetsCache()
+    {
+        var memoryStream = new MemoryStream();
+        var properties = new[] {
+            SourceId, SourceType, BasePath, ContentRoot, RelativePathPattern, RelativePathFilter,
+            AssetKind, AssetMode, AssetRole, AssetMergeSource, AssetMergeBehavior, RelatedAsset,
+            AssetTraitName, AssetTraitValue, CopyToOutputDirectory, CopyToPublishDirectory,
+            FingerprintCandidates.ToString()
+        };
+        var propertiesHash = HashingUtils.ComputeHash(memoryStream, properties);
+
+        var patternMetadata = new[] { nameof(FingerprintPattern.Pattern), nameof(FingerprintPattern.Expression) };
+        var fingerprintPatternsHash = HashingUtils.ComputeHash(memoryStream, FingerprintPatterns ?? [], patternMetadata);
+
+        var propertyOverridesHash = HashingUtils.ComputeHash(memoryStream, PropertyOverrides, nameof(ITaskItem.GetMetadata));
+
+        var candidateAssetMetadata = new[] {
+            "FullPath", "RelativePath", "TargetPath", "Link", "ModifiedTime", nameof(StaticWebAsset.SourceId),
+            nameof(StaticWebAsset.SourceType), nameof(StaticWebAsset.BasePath), nameof(StaticWebAsset.ContentRoot),
+            nameof(StaticWebAsset.AssetKind), nameof(StaticWebAsset.AssetMode), nameof(StaticWebAsset.AssetRole),
+            nameof(StaticWebAsset.AssetMergeBehavior), nameof(StaticWebAsset.AssetMergeSource), nameof(StaticWebAsset.RelatedAsset),
+            nameof(StaticWebAsset.AssetTraitName), nameof(StaticWebAsset.AssetTraitValue), nameof(StaticWebAsset.Fingerprint),
+            nameof(StaticWebAsset.Integrity), nameof(StaticWebAsset.CopyToOutputDirectory), nameof(StaticWebAsset.CopyToPublishDirectory),
+            nameof(StaticWebAsset.OriginalItemSpec)
+        };
+        var inputHashes = HashingUtils.ComputeHashLookup(memoryStream, CandidateAssets ?? [], candidateAssetMetadata);
+
+        var assetsCache = DefineStaticWebAssetsCache.ReadOrCreateCache(Log, CacheManifestPath);
+        assetsCache.PrepareForProcessing(propertiesHash, fingerprintPatternsHash, propertyOverridesHash, inputHashes);
+
+        return assetsCache;
+    }
+
+    internal class DefineStaticWebAssetsCache
+    {
+        private readonly List<ITaskItem> _assets = [];
+        private readonly List<ITaskItem> _copyCandidates = [];
+        private string? _manifestPath;
+        private IDictionary<string, ITaskItem> _inputHashes = new Dictionary<string, ITaskItem>();
```

Comment:
Nit: `_inputsByHash`? I thought this was a backing field for the property with the same name at first.